t/re-split.t

```diff
@@ -0,0 +1,707 @@
+                return
+            end
+
+            for i = 1, #my_table do
+                ngx.say(res[i])
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+a
+,
+b
+,
+c,d,e
+--- error_log eval
+qr/\[TRACE   \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 16: split matches, ctx arg
+--- http_config eval: $::HttpConfig
+--- config
+    location = /re {
+        content_by_lua_block {
+            local ngx_re = require "ngx.re"
+
+            local res, err = ngx_re.split("a,b,c,d,e", ",", nil, { pos = 5 })
+            if err then
+                ngx.log(ngx.ERR, "failed: ", err)
+                return
+            end
+
+            for i = 1, #res do
+                ngx.say(res[i])
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+c
+d
+e
+--- error_log eval
+qr/\[TRACE   \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 17: split matches, trailing subjects
+--- http_config eval: $::HttpConfig
+--- config
+    location /re {
+        content_by_lua_block {
+            local ngx_re = require "ngx.re"
+
+            local res, err = ngx_re.split(",a,b,c,d,", ",")
+            if err then
+                ngx.log(ngx.ERR, "failed: ", err)
+                return
+            end
+
+            for i = 1, #res do
+                if res[i] == "" then
+                    ngx.say("_blank_")
+                else
+                    ngx.say(res[i])
+                end
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+_blank_
+a
+b
+c
+d
+_blank_
+--- error_log eval
+qr/\[TRACE   \d+/
+--- no_error_log
+[error]
+attempt to get length of local 'regex' (a number value)
+
+
+
+=== TEST 18: split matches, real use-case
+--- http_config eval: $::HttpConfig
+--- config
+    location /re {
+        content_by_lua_block {
+            local ngx_re = require "ngx.re"
+
+            local res, err = ngx_re.split("abcd,erfg,ghij;hello world;aaa", ",|;")
+            if err then
+                ngx.log(ngx.ERR, "failed: ", err)
+                return
+            end
+
+            for i = 1, #res do
+                ngx.say(res[i])
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+abcd
+erfg
+ghij
+hello world
+aaa
+--- error_log eval
+qr/\[TRACE   \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 19: split no matches
+--- http_config eval: $::HttpConfig
+--- config
+    location /re {
+        content_by_lua_block {
+            local ngx_re = require "ngx.re"
+
+            local res, err = ngx_re.split("abcd", ",")
+            if err then
+                ngx.log(ngx.ERR, "failed: ", err)
+                return
+            end
+
+            for i = 1, #res do
+                ngx.say(res[i])
+            end
+        }
+    }
+--- request
+GET /re
+--- response_body
+abcd
+--- error_log eval
+qr/\[TRACE   \d+/
+--- no_error_log
+[error]
+
+
+
+=== TEST 20: subject is not a string type
+--- http_config eval: $::HttpConfig
```

Comment:
@thibaultcha Okay, I've digged into the valgrind error for you :) It's worth recording my process here:

1. I run TEST 3 in t/re-split separately by adding a `--- ONLY` section to that test block.
2. Also, I set the environment `export TEST_NGINX_USE_VALGRIND='--db-attach=yes --quiet --vgdb-error=1'` so that valgrind stops at the first memory issue and lets us use gdb to inspect the nginx process state. Additionally, I add the `---timeout: 1000` section to that test block too so that the client will not time out too soon.
3. By running the `re-split.t` test file with `prove`, it indeed stops at the first error:

    ```
t/re-split.t .. # I found ONLY: maybe you're debugging?
TEST 3: split matches, no submatch, jit compile, regex cache
==85110==
==85110== Warning: --db-attach is a deprecated feature which will be
==85110==    removed in the next release. Use --vgdb-error=1 instead
==85110==
==85110==
==85110== TO DEBUG THIS PROCESS USING GDB: start GDB like this
==85110==   /path/to/gdb nginx
==85110== and then give GDB the following command
==85110==   target remote | /usr/lib64/valgrind/../../bin/vgdb --pid=85110
==85110== --pid is optional if only one valgrind process is running
==85110==
==85110== Invalid read of size 16
==85110==    at 0x402407B: ???
==85110==    by 0x9054EBF: ???
==85110==    by 0xFFEFF646F: ???
==85110==  Address 0x9054ec0 is 16 bytes inside a block of size 24 alloc'd
==85110==    at 0x4C28C50: malloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==85110==    by 0x4C2AB1E: realloc (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==85110==    by 0x54690BF: lj_mem_realloc (lj_gc.c:811)
==85110==    by 0x546AE40: lj_str_new (lj_str.c:188)
==85110==    by 0x547E392: lj_parse_keepstr (lj_parse.c:237)
==85110==    by 0x547B799: lex_string (lj_lex.c:280)
==85110==    by 0x547B799: lex_scan (lj_lex.c:364)
==85110==    by 0x547BA1C: lj_lex_next (lj_lex.c:454)
==85110==    by 0x5480046: parse_args (lj_parse.c:1896)
==85110==    by 0x54801E2: expr_primary (lj_parse.c:1963)
==85110==    by 0x547EC2B: expr_simple (lj_parse.c:2009)
==85110==    by 0x547EC2B: expr_unop (lj_parse.c:2075)
==85110==    by 0x547EC2B: expr_binop (lj_parse.c:2088)
==85110==    by 0x547FEF6: expr (lj_parse.c:2107)
==85110==    by 0x547FEF6: expr_list (lj_parse.c:1874)
==85110==    by 0x54811BF: parse_local (lj_parse.c:2259)
==85110==    by 0x54811BF: parse_stmt (lj_parse.c:2646)
==85110==    by 0x54811BF: parse_chunk (lj_parse.c:2682)
==85110==
==85110== (action on error) vgdb me ...
```
4. Following the instructions provided in the valgrind output above, we use gdb to connect to valgrind:
    ```
gdb work/nginx/sbin/nginx
(gdb) lbt
builtin#169
@/home/agentzh/git/lua-resty-core/lib/ngx/re.lua:39
@/home/agentzh/git/lua-resty-core/lib/ngx/re.lua:165
=content_by_lua(nginx.conf:71):4
```

    Okay, so the Lua backtrace is shown above for the first memory error. Nice. BTW, I have the following `~/.gdbinit` file:

    ```gdb
directory /home/agentzh/git/openresty-gdb-utils

py import sys
py sys.path.append("/home/agentzh/git/openresty-gdb-utils")

source ngx-lua.gdb
source luajit21.py
set python print-stack full
```

    You may need to edit the file/directory paths above for your local file system layout accordingly.

    It's worth mentioning that we can find out what `builtin#169` is with the following command:

    ```
$ luajit $HOME/git/nginx-devel-utils/ljff.lua 169
FastFunc ffi.meta.__call
```

     Just ensure `luajit` is of exactly the same version of the one your nginx is currently loading. This is because different LuaJIT versions may have incompatible builtin Lua function IDs. So we can see that it's FFI's meta __call builtin.